### PR TITLE
Switch to using the Java HTTP client

### PR DESCRIPTION
**Pull request by melix**
There are 2 reasons to make that change:

1. reduce the number of dependencies of the client, in order to remove the dependency on Netty
2. the JDK HTTP client doesn't need to be closed

The 2d point deserves a bit more information. The test resources client factory is called by the test resources client, and in general, there should be at most a couple calls per build. In practice, this isn't the case: in Micronaut Data for example, there are many calls because tests spawn their own application context.

One issue is that Micronaut Core will _not_ call `#close` on a `PropertySourceLoader` once an application context is closed, even if that loader implements `Closeable`. It is arguable if it's a bug or not, but in practice, it means that there is no way for Micronaut Test Resources to proactively close a client.

Switching to the JDK HTTP Client solves this problem because there is no need to call the close method anymore.

This PR doesn't change the communication protocol between the client and the server because core lacks some support to make it easily possible. The consequence is that we're still adding a dependency on `micronaut-json-core`, which also means that if there's no JSON mapper on classpath (either Jackson or Serde), then test resources won't work. I am unsure but I think that was already the case before, even with the dependency on HTTP Netty, unless that was bringing a mapper on classpath transparently.

### CODE REVIEW

**Comment by melix** [comment]
@graemerocher @yawkat I _think_ there's a bug somewhere in our JSON handling. The fix I had to do in 961be32 looks pretty wrong.

In particular, on the _client project_, I am [mocking the server](https://github.com/micronaut-projects/micronaut-test-resources/blob/0cf995dca6b88363473f117497c37c6dde33930b/test-resources-client/src/test/groovy/io/micronaut/testresources/client/TestServer.groovy#L33). If that server returns a `String`, we get an answer like this:

```
HTTP/1.1 200 OK
Content-Type: application/json
connection: keep-alive
content-length: 5
date: Fri, 12 May 2023 07:26:03 GMT

world
```

Note that `Content-Type` is `application/json` as expected, but the string is returned as a raw string, which makes the json mapper scream:

> io.micronaut.json.JsonSyntaxException: Unrecognized token 'value': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')

This behavior isn't new, that's what is currently returned by the server in case the return type is `Optional<String>` (and I suspect the same with a plain `String` but I didn't check).

My fix was therefore to special case the `Argument.STRING` case on the client, so that it gets the raw string.
Now what gets even more interesting is what happens on the _server_ side, where in this case [I'm mocking the client with a declarative client](https://github.com/micronaut-projects/micronaut-test-resources/blob/7c1d05d0e2390de878e38e7147f351fb690a4490/test-resources-server/src/test/groovy/io/micronaut/testresources/server/TestResourcesControllerTest.groovy#L42).

If I'm using `testImplementation(mn.micronaut.http.client)` (which is actually the Netty declarative client), then my tests pass. But as soon as I switch to `testImplementation(mn.micronaut.http.client.jdk)` (the JDK http client), then my tests start failing with the very same error I faced in my own client implementation. So it seems that our implementation of the JDK client makes the same assumptions as I do, which is that the json mapper should be capable of handling the HTTP server response directly. Obviously that's not the case, and it seems that the Netty client does something special to workaround this issue. But I _think_ it means that the declarative http clients using the JDK client are broken.

Long story short, I don't understand what's going on.

**Comment by yawkat** [comment]
Yes, types like String (and byte[] and ByteBuffer) are treated specially and don't go through the json encoding/decoding process. It's possible that the jdk client is missing this special handling, I'm not familiar with it.

One solution could be to use the new MessageBodyHandler API. MessageBodyHandlerRegistry will handle this and return a "raw" decoder for String.

**Comment by melix** [comment]
So it means that my fix is correct, thanks!

**Review by graemerocher — APPROVED** [review]
